src/common/transformations/src/transformations/common_optimizations/align_mixed_fp32_fp16_types.cpp

```diff
@@ -0,0 +1,149 @@
+//
+// Created by pesir on 10/18/2022.
+//
+
+#include "transformations/common_optimizations/align_mixed_fp32_fp16_types.hpp"
+
+#include <openvino/opsets/opset1.hpp>
+#include <openvino/opsets/opset8.hpp>
+
+#include "itt.hpp"
+#include "openvino/op/util/precision_sensitive_attribute.hpp"
+#include "transformations/convert_precision.hpp"
+#include "transformations/rt_info/decompression.hpp"
+#include "transformations/rt_info/fp16_sensitive_subgraph_rt_info.hpp"
+
+using namespace ov;
+
+bool ov::pass::AlignMixedFP32FP16Types::run_on_model(const std::shared_ptr<ov::Model>& model) {
+    RUN_ON_MODEL_SCOPE(AlignMixedFP32FP16Types);
+
+    std::unordered_map<NodeTypeInfo, std::vector<size_t>> align_inputs_op_types_map{
+        {opset8::ReduceSum::get_type_info_static(), {0}},
+        {opset8::Exp::get_type_info_static(), {0}},
+        {op::util::BinaryElementwiseArithmetic::get_type_info_static(), {0, 1}},
+    };
+
+    std::function<bool(const std::shared_ptr<Node>&)> insert_converts_before = [&](const std::shared_ptr<Node>& node) {
+        bool is_changed = false;
+        bool is_matched_op_types_map = false;
+        for (auto const& item : align_inputs_op_types_map) {
+            if (!node->get_type_info().is_castable(item.first))
+                continue;
+            is_matched_op_types_map = true;
+
+            for (auto const& input_idx : item.second) {
+                auto incoming_output = node->input_value(input_idx);
+                if (is_fp16_precision_sensitive(incoming_output.get_node_shared_ptr()))
+                    continue;  // we are in the middle
+                const auto& existing_convert =
+                    std::dynamic_pointer_cast<opset8::Convert>(incoming_output.get_node_shared_ptr());
+                if (existing_convert) {
+                    if (existing_convert->get_convert_element_type() != element::f32) {
```

Comment:
check if this Convert is really needed?

Reply:
there is no harm from such a check, just 2 lines and if converts are already there we don't need to recreate them